### PR TITLE
Solve cntrl+f issue on KVv2 JSON details and edit views

### PR DESCRIPTION
**Pull request by Monkeychip**
- [x] ent tests pass

### Description
Address issue #28053 and closes PR #28604.

1. Replace the codemirror component on the KVv2 details page with the `Hds::CodeBlock` component. This solves the original issue, allowing the user to browser search the full secret while also moving us closer to HDS adoption.

2. For the edit view, this problem is solved by correctly taking advantage of the `viewportMargin` param provided by [CodeMirror](https://codemirror.net/5/doc/manual.html).

image
 
**Notes:**
- For KVv2 only I calculate the `viewportMargin` based on the number of JSON object keys. I use the higher of the two numbers 10 or key count, but no higher than 1,000. Before we would also be using 10 (the default).

- For the use cases where we were passing in "Infinity" this has negative performance implications and the only upside to ~~us~~ me three years ago misspelling `viewportMargin` 🫣. 

https://github.com/user-attachments/assets/7fc95d55-ec1c-475b-8c31-47ac0df8b9bd

### CODE REVIEW

**Comment by karabanov** [comment]
I agree that this function may not be needed by everyone, but maybe it is possible to somehow make it permanently enabled for those who need it? That is, so that when you press this button, its state is remembered, for example, somewhere in the LocalStorage and remains such that you do not have to press the button every time

**Comment by Monkeychip** [comment]
@karabanov thank you so much for the feedback. We're running into an issue with adding this functionality to our current component and are looking at replace it with a [HDS::codeblock](https://hds-website-hashicorp.vercel.app/components/code-block?tab=code#how-to-use-this-component) component. The HDS codeblock allows you to `cntrl + f` inside a fixed height container so that you can search a whole secret. Would this solve you're problem, or is there still a desire to see the whole secret with the expand button? See screenshare of the new component below.

Re: localStorage, that's a potential option, but in general we try to minimize our use of localStorage. We've had similar requests for things like pinning an item to the top of a list, and other config-type requests. I'll throw this by the team, but no promises.

https://github.com/user-attachments/assets/5e3724b3-2a3f-4afc-b66a-9b4a6d40fc69

**Comment by karabanov** [comment]
@Monkeychip, I think if the `ctrl+f` search works correctly, then my colleagues and I will be quite satisfied.

Please tell me, will the search work in edit mode if you use the [HDS::codeblock](https://hds-website-hashicorp.vercel.app/components/code-block?tab=code#how-to-use-this-component) component? The fact is that if you stretch the codemirror window, the search will work both in viewing and editing mode

**Comment by Monkeychip** [comment]
@karabanov the `Hds::Codeblock` would only be for the details view not the edit as it's not designed to be a code editor. So this would solve the details view but not edit. I can make/push this PR to address the details view but for edit, that is a longer process and will involved doing some modifications to our current code editor (if possible). I'll go ahead and make a ticket for it and see if it's something we can address in another pr.

**Comment by karabanov** [comment]
@Monkeychip A working search by `ctrl+f` is also needed in the editing mode, since the JSON can be large and it is problematic to find the necessary key in it with your eyes to edit it, in addition, the keys are sorted in an unnatural order, which complicates the task even more

For example, if you do not switch to the JSON editing mode, all the keys will be visible on the page at once, they can be found by searching both in the viewing mode and in the editing mode, so why not just stretch the text field by 100% in the JSON editing mode so that all the keys become visible? This will be very similar to the normal editing mode, when all the keys are visible and the problem with searching will be solved.

The improvement that I propose in PR #28604 is to add one line and this will not require replacing the text editor, while solving the problem with searching for keys and displaying the content. This will even make the JSON editing mode more similar to the normal mode, since all the keys will be visible at once

How do you like this solution?

**Comment by Monkeychip** [comment]
@karabanov  All very fair points. I went back to debugging why our current code editor does not allow searching beyond the viewport and got to the heart of the problem. We were incorrectly using a param called `viewportMargin`. This param tells the json editor to render the number of lines beyond what is visible in the viewport. **A quick summary**: it was defaulting to rendering 10 lines beyond the viewport, which for longer secrets is not enough. I amended this so the `viewportMargin` is dynamically calculated based on the size of the secret. 

In conclusion:
* Short term fix: Replace **details** view with new `Hds::CodeBlock` component that allows search. Update **edit** view to dynamical render the full secret which will effectively allow you to search the whole secret. No expand button.

* Longer term fix: We're working on getting a new code editor that will solve this problem and possibly include an expand button for the edit and readOnly views because you make a good argument for it and I put a request into the design team —thank you!

**Comment by karabanov** [comment]
@Monkeychip Thank you very much
I look forward to these changes being released soon

File: ui/app/templates/components/console/log-json.hbs [hunk 1]
```diff
@@ -8,7 +8,8 @@
     @showToolbar={{false}}
     @value={{stringify this.content}}
     @readOnly={{true}}
-    @viewportMargin="Infinity"
+    {{! ideally we calculate the "height" of the json data, but 100 should cover most cases }}
```
Comment (Monkeychip) [inline ui/app/templates/components/console/log-json.hbs:11]: previously this was defaulting to 10, so another option is to remove this all together and let it default to 10 until someone complains?

File: ui/app/templates/components/control-group-success.hbs [hunk 2]
```diff
@@ -21,7 +21,8 @@
           @showToolbar={{false}}
           @value={{stringify this.unwrapData}}
           @readOnly={{true}}
-          @viewportMargin="Infinity"
+          {{! ideally we calculate the "height" of the json data, but 100 should cover most cases }}
```
Comment (Monkeychip) [inline ui/app/templates/components/control-group-success.hbs:24]: same comment as above.

File: ui/lib/kv/addon/components/kv-patch/subkeys-reveal.hbs [hunk 3]
```diff
@@ -8,6 +8,6 @@
     <Hds::Text::Body @tag="p" @weight="semibold">Reveal subkeys in JSON</Hds::Text::Body>
   </Toggle>
   {{#if this.showSubkeys}}
-    <Hds::CodeBlock @value={{stringify @subkeys}} @hasLineNumbers={{false}} data-test-subkeys />
+    <Hds::CodeBlock @value={{stringify @subkeys}} @hasLineNumbers={{false}} data-test-code-block="subkeys" />
```
Comment (Monkeychip) [inline ui/lib/kv/addon/components/kv-patch/subkeys-reveal.hbs:11]: I updated all selectors added to the `Hds::CodeBlock` component so that we consistently used one GENERAL selector.

File: ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-edge-cases-test.js [hunk 4]
```diff
@@ -299,22 +299,34 @@ module('Acceptance | kv-v2 workflow | edge cases', function (hooks) {
...
+      'JSON editor displays correct empty object'
     );
     codemirror().setValue('{ "foo3": { "name": "bar3" } }');
     await click(FORM.saveBtn);
 
     // Details view
     await click(PAGE.secretTab('Secret'));
-    assert.dom(FORM.toggleJson).isNotDisabled();
-    assert.dom(FORM.toggleJson).isChecked();
-    assert.false(codemirror().getValue().includes('*'), 'Values are not obscured on details view');
```
Comment (Monkeychip) [inline ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-edge-cases-test.js:311]: A little clean up that was missed when we removed obscuring json values from the details view PR #28261.

File: ui/tests/helpers/kv/kv-selectors.js
Comment (Monkeychip) [inline ui/tests/helpers/kv/kv-selectors.js:1]: Going back and forth a bit with adding this to general selectors or keeping here as is. I am leaning towards as is and seeing how often this is going to be used in other places before making these GENERAL selectors adding to the GENERAL.codeBlock().

That being said, no strong opinions. Thoughts?

File: ui/lib/kv/addon/components/kv-data-fields.hbs [hunk 5]
```diff
@@ -13,12 +13,27 @@
...
+      @maxHeight="300px"
+      as |CB|
+    >
+      <CB.Title @tag="h3">
+        Version data
+      </CB.Title>
+    </Hds::CodeBlock>
+  {{else}}
+    <JsonEditor
+      @title="{{if (eq @type 'create') 'Secret' 'Version'}} data"
```
Comment (Monkeychip) [inline ui/lib/kv/addon/components/kv-data-fields.hbs:31]: Note: When you create a new version you see Version data vs Secret data.

File: ui/tests/helpers/kv/kv-selectors.js [hunk 6]
```diff
@@ -90,8 +90,8 @@ export const PAGE = {
   },
   paths: {
     copyButton: (label) => `${PAGE.infoRowValue(label)} button`,
```
Comment (Monkeychip) [inline ui/tests/helpers/kv/kv-selectors.js:92]: I go back and forth on putting this in the general selectors. Opted to keep as you see here until it becomes more obvious that these selectors will be used outside of kv.

File: ui/tests/helpers/secret-engine/secret-engine-helpers.js [hunk 7]
```diff
@@ -206,3 +206,49 @@ export const fillInAwsConfig = async (situation = 'withAccess') => {
     await fillIn(GENERAL.ttl.input('Identity token TTL'), '7200');
   }
 };
+
```
Comment (Monkeychip) [inline ui/tests/helpers/secret-engine/secret-engine-helpers.js:209]: Admittedly, I got carried away on making this helper. I could have saved myself time and hardcoded or modified the KvDataFields defaults during a test situation. However, I can imagine this being a helpful function for future KV things. Gives us a quick way to make longer more complicated json objects. Also, it was kind of fun.

**Review by hellobontempo — COMMENTED** [review]
Great work on this! Really thorough and nice communication with the reporter to get this resolved the right way. 

Mostly just minor cleanup comments and questions. I'm curious about the accessibility change. It's end of day for me so feeling brain dead. Can circle back with an approval tomorrow ☺️

File: ui/app/templates/components/console/log-json.hbs (hunk 1 above)
Comment (Monkeychip) [reply ui/app/templates/components/console/log-json.hbs:11]: From what I can gather this is output for the ui console/terminal window. 10 for most situations should be enough unless they're trying to read a really long secret. 

Is 100 going to have any performance implications? not that I can tell via testing, but also there's no documentation to say otherwise. From what I gather, they're cautioning against using "Infinity" because that could be bad if your json was millions of lines long.

File: ui/lib/core/addon/components/json-editor.hbs [hunk 8]
```diff
@@ -51,7 +51,7 @@
       mode=@mode
       readOnly=@readOnly
       theme=@theme
-      viewportMarg=@viewportMargin
+      viewportMargin=@viewportMargin
```
[the remark this answers is not in the document]
Comment (Monkeychip) [reply ui/lib/core/addon/components/json-editor.hbs:54]: The default is inherit from CodeMirror. If you don't pass `viewportMargin` it will default to 10, which was it was doing when I incorrectly spelled the param.

File: ui/lib/core/addon/components/json-editor.js [hunk 9]
```diff
@@ -24,7 +24,7 @@ import { action } from '@ember/object';
  * @param {Boolean} [readOnly] - Sets the view to readOnly, allowing for copying but no editing. It also hides the cursor. Defaults to false.
  * @param {String} [theme] - Specify or customize the look via a named "theme" class in scss.
  * @param {String} [value] - Value within the display. Generally, a json string.
- * @param {String} [viewportMargin] - Size of viewport. Often set to "Infinity" to load/show all text regardless of length.
+ * @param {String} [viewportMargin] - Specifies the amount of lines that are rendered above and below the part of the document that's currently scrolled into view. Determines how much of the json object is rendered and what in the object can be searched using cntrl + f.
```
[the remark this answers is not in the document]
Comment (Monkeychip) [reply ui/lib/core/addon/components/json-editor.js:27]: I pulled the first chunk from documentation, but yes I think this suffices.

File: ui/tests/integration/components/json-editor-test.js [hunk 10]
```diff
@@ -29,13 +30,16 @@ module('Integration | Component | json-editor', function (hooks) {
...
+    this.set('long_json', JSON.stringify(createLongJson(), null, `\t`));
     this.set('hashi-read-only-theme', 'hashi-read-only auto-height');
     setRunOptions({
       rules: {
         // CodeMirror has a secret textarea without a label that causes this problem
         label: { enabled: false },
         // TODO: investigate and fix Codemirror styling
         'color-contrast': { enabled: false },
+        // failing on .CodeMirror-scroll
+        'scrollable-region-focusable': { enabled: false },
```
[the remark this answers is not in the document]
Comment (Monkeychip) [reply ui/tests/integration/components/json-editor-test.js:42]: Good question. This is an ally testing thing—copied the same thing from Chelsea's fixes, here's [one](https://github.com/hashicorp/vault/blob/main/ui/tests/integration/components/database-role-setting-form-test.js#L108) example but there are several more. From what I can tell it's because I'm testing just the component and now it scrolls.

File: ui/lib/core/addon/components/json-editor.hbs (hunk 8 above)
[the remark this answers is not in the document]
Comment (hellobontempo) [reply ui/lib/core/addon/components/json-editor.hbs:54]: Oh - is that default just builtin? In codemirror I see a default of an empty string ([here](https://github.com/hashicorp/vault/blob/main/ui/lib/core/addon/modifiers/code-mirror.js#L79)) Maybe we could include that in the JSDOC below that it defaults to 10?

**Review by hellobontempo — APPROVED** [review]
Great work! Just a comment about mentioning the default `viewportMargin` in the JSON editor.
